File: scripts/ColorPublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import ColorCommand

class ColorPublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", ColorCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", ColorCommand, queue_size=10)

        while not rospy.is_shutdown():

            self.message = ColorCommand()

            self.message.command = "OFF"
            self.message.iscommand = False
            self.message.ispercentage = False
            self.message.percentage = 0
            self.message.ishsb = True
            self.message.hue = 127
            self.message.saturation = 61
            self.message.brightness = 99
            
            self.message.isnull = False

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: command = %s, iscommand = %s, ispercentage = %s, percentage = %s, ishsb = %s, hue = %s, saturation = %s, brightness = %s" % (self.message.item, rospy.get_time(), self.message.command, self.message.iscommand, self.message.ispercentage, self.message.percentage, self.message.ishsb, self.message.hue, self.message.saturation, self.message.brightness)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("ColorPublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    colorPublisher = ColorPublisher("testColor")

    try:
        colorPublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/ColorPublisher.cpp
#include <ros/ros.h>
#include <openhab_msgs/ColorCommand.h>

class ColorPublisher
{
public:
    ColorPublisher(std::string item_name)
    {
        item_name_ = item_name;
        pub_ = nh_.advertise<openhab_msgs::ColorCommand>("/openhab/items/" + item_name_ + "/command", 10);
        rate_ = ros::Rate(10); // 10Hz
        enable_ = false;
        message_ = NULL;

        if (enable_)
        {
            start();
        }
        else
        {
            stop();
        }
    }

    void start()
    {
        enable_ = true;
        pub_ = nh_.advertise<openhab_msgs::ColorCommand>("/openhab/items/" + item_name_ + "/command", 10);

        while (ros::ok())
        {
            message_.command = "OFF";
            message_.iscommand = false;
            message_.ispercentage = false;
            message_.percentage = 0;
            message_.ishsb = true;
            message_.hue = 127;
            message_.saturation = 61;
            message_.brightness = 99;
            message_.isnull = false;

            message_.header.stamp = ros::Time::now();
            message_.header.frame_id = "/base_link";
            message_.item = item_name_;

            std::string message = "Publishing to " + message_.item + " at " + std::to_string(ros::Time::now().toSec()) + ": command = " + message_.command + ", iscommand = " + std::to_string(message_.iscommand) + ", ispercentage = " + std::to_string(message_.ispercentage) + ", percentage = " + std::to_string(message_.percentage) + ", ishsb = " + std::to_string(message_.ishsb) + ", hue = " + std::to_string(message_.hue) + ", saturation = " + std::to_string(message_.saturation) + ", brightness = " + std::to_string(message_.brightness);
            ROS_INFO("%s", message.c_str());

            pub_.publish(message_);
            rate_.sleep();
        }
    }

    void stop()
    {
        enable_ = false;
        pub_.shutdown();
    }

private:
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Rate rate_;
    bool enable_;
    std::string item_name_;
    openhab_msgs::ColorCommand message_;
};

int main(int argc, char **argv)
{
    // Initialize the node and name it.
    ros::init(argc, argv, "ColorPublisherNode");

    // Create an object of class ColorPublisher that will take care of everything.
    ColorPublisher colorPublisher("testColor");

    try
    {
        colorPublisher.start();
    }
    catch (ros::Exception &e)
    {
        ROS_ERROR("Error occurred: %s", e.what());
    }

    // Allow ROS to go to all callbacks.
    ros::spin();
    
    return 0;
}
